### PR TITLE
Getting Started with ECLAT Algorithm in Association Rule Mining

### LINKED ISSUE #5161: Getting Started with ECLAT Algorithm in Association Rule Mining
*NOTE*: All **1st-time contributors** should know Topic approval and PR submission does NOT guarantee your Topic/Article will be published. Our team of Peer Reviewers and Content Moderators will review all PRs that come in to make sure they adhere to the standard of quality we expect from the EngEd community. 

We expect all community memebers to go through provided examples, resources, and previously published material before submitting content. As a rule of thumb, please only submit articles (pull requests) that are complete, formatted correctly, and include a fully polished article (ready to be published and error free). 

All revisions and edits should be *completed on your own forked repo (as to not take up room in the queue)*. **Any PR that is submitted incomplete (meaning not ready to be published as is) will be closed.** A PR being closed does NOT mean the article can not be published (or fixed) it simply means that the edits, revisions, and fixes will happen outside of the queue.

### Topic Suggestion
Enter topic suggestion here...

#### Pre-submission advice
By following all our pre-submission advice and reviewing our Resources folder, you will maximise your chances of your topic being approved.

We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner.

Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published (subject to vary depending on _volume and/or backlog of articles_).

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles that will help you propose and write a successful article.

**Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

### Writing sample(s): 
Include any links or writing samples - to help our team better gauge your writing quality.

#### Proposed article introduction

The ECLAT, which stands for Equivalence Class Transformation, is an association rule mining algorithm that extracts frequent itemsets from a database. Frequent itemsets are a set of items that are frequently bought together. Other methods used in association rule mining to obtain rules are Apriori and FP Growth algorithms. Both Apriori and FP Growth use horizontal data format, unlike Eclat which mines itemsets using the vertical data format, and it's a depth-first search algorithm.

Apriori was the first algorithm developed in this field to find association rules. However, this algorithm is not very efficient, and it is not suitable for large databases. ECLAT algorithm came in to solve this limitation of the Apriori algorithm, and indeed it's a scalable version of the Apriori algorithm. While the Apriori algorithm works horizontally, imitating the Breadth-First Search of a graph, the ECLAT algorithm works vertically just like the Depth-First Search of a graph. This vertical approach of the ECLAT algorithm makes it a faster algorithm than the Apriori algorithm.

The basic idea is to use Transaction Identifier Sets(TID) intersections to compute a candidate's support value and avoid the generation of subsets that do not exist in the prefix tree. In the first call of the function, all single items are used along with their TID. Then the function is called recursively, and in each recursive call, each item-	TID pair is verified and combined with other item-TID pairs. This process is continued until no candidate item-TID pairs can be combined.

In this article, we shall learn everything we need to know to get started with the Eclat algorithm.
Using an arbitrary example, we shall learn how this algorithm work. Later we will discuss how this algorithm differs from the Apriori algorithm and move a step ahead to discuss its advantages in association learning. Finally, we shall implement this algorithm in python.

#### Key takeaways 
By the end of this tutorial, a learner is expected to know;
- What is Association rule learning
- What is the ECLAT algorithm
- How Eclat algorithm work
- Difference between ECLAT and Apriori algorithm
- Advantages of ECLAT algorithm
- How to implement ECLAT algorithm in python
#### Article quality
This article will provide a deep explanation of the ECLAT algorithm using a visual example where all calculations will be done step-by-step.

#### References
N/A
### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by jacksonjate1**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5161

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by Bayler** [comment]
@jacksonjate1 It is good practice to always provide a Google Colab link for your implementation for reproducibility. Please include one.

**Comment by Kayere** [comment]
Plagiarism check done.

**Comment by jacksonjate1** [comment]
@Bayler 
I'll be adding the link to the source code as fast as possible, that's, within 24 hours

**Comment by jacksonjate1** [comment]
@Bayler 
I've added the link to the source code.

File: content/articles/eclat-algorithm-in-python/index.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+```
+Now that our output is organized into a pandas data frame, we can have a look at the first seven supports as follows;
+
+```python
+rslt_DataFrame.nlargest(n = 7, columns = 'Support') # printing the first 7 supports
+
+```
+![Arranged results](/engineering-education/eclat-algorithm-in-python/supports.png)
+ 
+From this output, we notice that the support of the first rule(herb & pepper with) is the highest, followed by the (whole wheat pasta with olive oil) ans so on. This table means that the first rule is the most important. Therefore, for the seller to maximize their sales and profit, they should follow the first rule.
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:143]: I'd like you to re-read this paragraph and correct the grammatical mistakes accordingly.

File: content/articles/eclat-algorithm-in-python/index.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+To have the best rules, we shall adopt the Apriori algorithm in our implementation. To get started, we need to import the necessary libraries for this session.
+
+```python
+# The first thing is to install this package
+!pip install apyori
+
+```
+### Data Preprocessing
+#### Importing the libraries
+The libraries we shall work with in this session are as follows;
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:94]: Avoid using `shall` in your tutorial. Always use `will` or an alternative word.

File: content/articles/eclat-algorithm-in-python/index.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+2. Intersect tidlist of item {a} with tidlist of item {b} and generate a new tansaction list that both items {a} and {b} are found.
+3. Reapet step 1 on {a}-coditional database.
+4. For all other items, repeat the above steps.
+
+If we perform these steps on the verticle dataset we gave above, we should obtain an output similar to the one below;
+![data image](/engineering-education/eclat-algorithm-in-python/tid-set.png)
+
+
+### Advantages
+1. Eclat algorithm has low memory requirements compared to Apriori as it uses a Depth-First Search approach.
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:74]: You can either explain what "Depth-First Search" approach is or add an external link to it.

File: content/articles/eclat-algorithm-in-python/index.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+### Eclat Algorithm
+1. Get tidlist for each item.
+ Here we need to scan the entire database. The tidlist of item {a} is the list of all transactions in which item {a} is found.
+2. Intersect tidlist of item {a} with tidlist of item {b} and generate a new tansaction list that both items {a} and {b} are found.
+3. Reapet step 1 on {a}-coditional database.
+4. For all other items, repeat the above steps.
+
+If we perform these steps on the verticle dataset we gave above, we should obtain an output similar to the one below;
+![data image](/engineering-education/eclat-algorithm-in-python/tid-set.png)
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:70]: Are these images yours? Or you got them from other sources?

File: content/articles/eclat-algorithm-in-python/index.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+We obtain a verticle data format by making a list of transactions to which each particular item is found.
+
+Let's look at the steps in the Eclat algorithm.
+
+### Eclat Algorithm
+1. Get tidlist for each item.
+ Here we need to scan the entire database. The tidlist of item {a} is the list of all transactions in which item {a} is found.
+2. Intersect tidlist of item {a} with tidlist of item {b} and generate a new tansaction list that both items {a} and {b} are found.
+3. Reapet step 1 on {a}-coditional database.
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:66]: Please proofread your work.

File: content/articles/eclat-algorithm-in-python/index.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+2. Eclat Algorithm
+3. FP Growth Algorithm
+
+The first algorithm to be introduced in the Data Mining domain was the Apriori algorithm. This algorithm, however, had some limitations in discovering frequent itemsets. Its limitations created a need for a more efficient algorithm.
+
+Later, the Eclat algorithm was introduced to deal with the weakness of the Apriori algorithm. Between these two algorithms, we have a significant difference.
+
+The first difference is that, unlike the Apriori algorithm, which is applicable with *horizontal dataset*, the Eclat algorithm is applicable only with a dataset in *verticle dataset* format.
+
+Secondly, in the Eclat algorithm, only the *support* is counted, and the confidence, as in the case of Apriori, is not computed. Here, the *Support* is nothing but the number of times an item is in a database.
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:42]:
```suggestion
Later, the Eclat algorithm was introduced to deal with the weakness of the Apriori algorithm. Between these two algorithms, we have significant differences:
- Unlike the Apriori algorithm, which is applicable with *horizontal dataset*, the Eclat algorithm is applicable only with a dataset in *verticle dataset* format.
- In the Eclat algorithm, only the *support* is counted, and the confidence, as in the case of Apriori, is not computed. Here, the *Support* is nothing but the number of times an item is in a database.
```

File: content/articles/eclat-algorithm-in-python/index.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+The key takeaway here is that Eclat works with a vertical data format. Since most datasets are in horizontal format, to apply the Eclat algorithm, we first have to convert them to vertical format.
+
+Below are examples of Horizontal and verticle data formats;
+
+![Horizontal data](/engineering-education/eclat-algorithm-in-python/horizontal.png)
+Horizontal data format
+
+![Verticle data](/engineering-education/eclat-algorithm-in-python/verticle-format.png)
+Verticle data format
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:56]: Please confirm to me that the image names correspond to their respective images.

**Comment by Bayler** [comment]
@jacksonjate1 Please work on the revisions before we can move on for final review.

File: content/articles/eclat-algorithm-in-python/index.md (hunk 7 above)
Comment (jacksonjate1) [reply content/articles/eclat-algorithm-in-python/index.md:56]: They are similar.

File: content/articles/eclat-algorithm-in-python/index.md (hunk 4 above)
Comment (jacksonjate1) [reply content/articles/eclat-algorithm-in-python/index.md:70]: The image is mine.

**Comment by jacksonjate1** [comment]
@Bayler 
Am done with the requested revision

File: content/articles/eclat-algorithm-in-python/index.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+Since we are all well with the intuitions, let's proceed and implement this algorithm in python.
+
+### Python Implementation of the Eclat Algorithm
+To have the best rules, we shall adopt the Apriori algorithm in our implementation. To get started, we need to import the necessary libraries for this session.
+
+```python
+# The first thing is to install this package
+!pip install apyori
+
+```
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:91]: This isn't a python command but rather a bash command

File: content/articles/eclat-algorithm-in-python/index.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+### Prerequisites:
+To follow up with this article, the learner needs to be comfortable working with the following libraries in Python: Scikit-learn, Pandas, and Numpy.
+
+### Introduction to ECLAT Algorithm
+As we aforemebtioned, the main idea in the association rule is to discover valid information and knowledge from a large dataset. There are a number of algorithms that have been developed over the years that make this activity as successful as possible. The major algorithm used includes:
+
+1. Apriori Algorithm
+2. Eclat Algorithm
+3. FP Growth Algorithm
```
Comment (Bayler) [inline content/articles/eclat-algorithm-in-python/index.md:34]: @jacksonjate1 Since you've mentioned the `FP Growth Algorithm`, briefly talk about it.

**Comment by Bayler** [comment]
@jacksonjate1 Two more edits before we can move on for final review.

**Comment by LinusMuema** [comment]
Hi @jacksonjate1 

Awesome job on the article. Please take note of the changes made during the review process.

cc: @Bayler

**Pull request merged**